In [1]:
import numpy as np 
import urdf2casadi.urdfparser as u2c
from urdf2casadi.geometry import plucker
from urdf_parser_py.urdf import URDF, Pose
import PyKDL as kdl
import kdl_parser_py.urdf as kdlurdf
from timeit import Timer, timeit, repeat
import rbdl
import pybullet as pb


In [2]:
def median(lst):
    n = len(lst)
    if n < 1:
            return None
    if n % 2 == 1:
            return sorted(lst)[n//2]
    else:
            return sum(sorted(lst)[n//2-1:n//2+1])/2.0
        
def average(lst): 
    return sum(lst) / len(lst) 


def c_kdl_func():
    for j in range(njoints):
        q_kdl[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_kdl[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    kdl.ChainDynParam(kdlmodel, gravity_kdl).JntToCoriolis(q_kdl, qdot_kdl, g_kdl)


def c_u2c_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    c_u2c(q_none, qdot_none)
    
def c_pb_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    c_pb = pb.calculateInverseDynamics(pbmodel, q_none, qdot_none, qddot)
    
    
def c_rbdl_func():
    for j in range(njoints):
        q_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    rbdl.NonlinearEffects(rbdlmodel, q_np, qdot_np, c_rbdl)


def g_kdl_func():
    for j in range(njoints):
        q_kdl[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    kdl.ChainDynParam(kdlmodel, gravity_kdl).JntToGravity(q_kdl, g_kdl)


def g_u2c_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    g_u2c(q_none)
    
def g_pb_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    g_pb = pb.calculateInverseDynamics(pbmodel, q_none, qdot, qddot)
    
def g_rbdl_func():
    for j in range(njoints):
        q_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    rbdl.InverseDynamics(rbdlmodel, q_np, qdot_np, qddot_np, g_rbdl)
    
    
def ID_u2c_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    ID_u2c(q_none, qdot_none, qddot_none)
    
    
def ID_pb_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2 
    ID_pb = pb.calculateInverseDynamics(pbmodel, q_none, qdot_none, qddot_none)
    
    
def ID_rbdl_func():
    for j in range(njoints):
        q_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    rbdl.InverseDynamics(rbdlmodel, q_np, qdot_np, qddot_np, ID_rbdl)

    
def M_kdl_func():
    for j in range(njoints):
        q_kdl[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    kdl.ChainDynParam(kdlmodel, gravity_kdl).JntToMass(q_kdl, M_kdl)


def M_u2c_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    M_u2c(q_none)
    
def M_pb_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    H = pb.calculateMassMatrix(pbmodel, q_none)
    
    
def M_rbdl_func():
    for j in range(njoints):
        q_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    rbdl.CompositeRigidBodyAlgorithm(rbdlmodel, q_np, M_rbdl)
    
    
def aba_u2c_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    aba_u2c(q_none, qdot_none, tau_none)
    
    
def crba_u2c_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    crba_u2c(q_none, qdot_none, tau_none)
    
    
def fd_rbdl_func():
    for j in range(njoints):
        q_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    rbdl.ForwardDynamics(rbdlmodel, q_np, qdot_np, tau_np, qddot_np)
    

In [3]:
#pbmodel = pb.loadURDF("pantilt.urdf")
sim = pb.connect(pb.DIRECT)
#pbmodel = pb.loadURDF("/urdf4timing/1dof.urdf")
ndofs = 10
urdf_nr = list(range(ndofs + 1))
urdf_nr.pop(0)


nitr = 1000
nrepeat = 100
    

path_to_urdf = "/home/lmjohann/urdf2casadi/examples/urdf/pantilt.urdf"
root = 'base_link'
tip = "tilt_link"

ok, kdl_tree = kdlurdf.treeFromFile(path_to_urdf)
kdlmodel = kdl_tree.getChain(root,tip)

rbdlmodel = rbdl.loadModel(path_to_urdf)
pbmodel = pb.loadURDF(path_to_urdf, useFixedBase=True, flags = pb.URDF_USE_INERTIA_FROM_FILE)
asd = u2c.URDFparser()
robot_desc = asd.from_file(path_to_urdf)

jlist, names, qmax, qmin = asd.get_joint_info(root, tip)
njoints = asd.get_n_joints(root, tip)

gravity_kdl = kdl.Vector()
gravity_kdl[2] = -9.81
gravity_u2c = [0, 0, -9.81]

#rbdl declarations
q_np = np.zeros(njoints)
qdot_np = np.zeros(njoints)
qddot_np = np.zeros(njoints)
g_rbdl = np.zeros(njoints)
c_rbdl = np.zeros(njoints)
ID_rbdl = np.zeros(njoints)

#kdl declatations
q_kdl = kdl.JntArray(njoints)
qdot_kdl = kdl.JntArray(njoints)
g_kdl = kdl.JntArray(njoints)

#u2c and pybullet declarations
q_none = [None]*njoints
qdot_none = [None]*njoints
qddot_none = [None]*njoints
g_pb = [None]*njoints

qddot = []
for i in range(njoints):
    qddot.append(0.)

qdot = []
for i in range(njoints):
    qdot.append(0.)
    
    
M_kdl = kdl.JntSpaceInertiaMatrix(njoints)
M_rbdl = (rbdlmodel.q_size, rbdlmodel.q_size)
M_rbdl = np.zeros(M_rbdl)
tau_np = np.zeros(njoints)
tau_none = [None]*njoints




The root link base_link has an inertia specified in the URDF, but KDL does not support a root link with an inertia.  As a workaround, you can add an extra dummy link to your URDF.
/home/lmjohann/urdf2casadi/examples/urdf/pantilt.urdf


In [4]:
g_u2c = asd.get_gravity_rnea(root, tip, gravity_u2c)
timeit_g_u2c = repeat("g_u2c_func()", setup = "from __main__ import g_u2c_func", repeat = nrepeat, number = nitr)
mediantime_g_u2c  = median(timeit_g_u2c)/nitr*1000000

timeit_g_pb = repeat("g_pb_func()", setup = "from __main__ import g_pb_func", repeat = nrepeat, number = nitr)
mediantime_g_pb = median(timeit_g_pb)/nitr*1000000

timeit_g_rbdl = repeat("g_rbdl_func()", setup = "from __main__ import g_rbdl_func", repeat = nrepeat, number = nitr)
mediantime_g_rbdl = median(timeit_g_rbdl)/nitr*1000000

timeit_g_kdl = repeat("g_kdl_func()", setup = "from __main__ import g_kdl_func", repeat = nrepeat, number = nitr)
mediantime_g_kdl = median(timeit_g_kdl)/nitr*1000000


c_u2c = asd.get_coriolis_rnea(root, tip)
timeit_c_u2c = repeat("c_u2c_func()", setup = "from __main__ import c_u2c_func", repeat = nrepeat, number = nitr)
mediantime_c_u2c  = median(timeit_c_u2c)/nitr*1000000

timeit_c_pb = repeat("c_pb_func()", setup = "from __main__ import c_pb_func", repeat = nrepeat, number = nitr)
mediantime_c_pb = median(timeit_c_pb)/nitr*1000000

timeit_c_rbdl = repeat("c_rbdl_func()", setup = "from __main__ import c_rbdl_func", repeat = nrepeat, number = nitr)
mediantime_c_rbdl = median(timeit_c_rbdl)/nitr*1000000

timeit_c_kdl = repeat("c_kdl_func()", setup = "from __main__ import c_kdl_func", repeat = nrepeat, number = nitr)
mediantime_c_kdl = median(timeit_c_kdl)/nitr*1000000

ID_u2c = asd.get_inverse_dynamics_rnea(root, tip, gravity_u2c)
timeit_id_u2c = repeat("ID_u2c_func()", setup = "from __main__ import ID_u2c_func", repeat = nrepeat, number = nitr)
mediantime_id_u2c  = median(timeit_id_u2c)/nitr*1000000

timeit_id_pb = repeat("ID_pb_func()", setup = "from __main__ import ID_pb_func", repeat = nrepeat, number = nitr)
mediantime_id_pb = median(timeit_id_pb)/nitr*1000000

timeit_id_rbdl = repeat("ID_rbdl_func()", setup = "from __main__ import ID_rbdl_func", repeat = nrepeat, number = nitr)
mediantime_id_rbdl = median(timeit_id_rbdl)/nitr*1000000


M_u2c = asd.get_inertia_matrix_crba(root, tip)
timeit_M_u2c = repeat("M_u2c_func()", setup = "from __main__ import M_u2c_func", repeat = nrepeat, number = nitr)
mediantime_M_u2c  = median(timeit_M_u2c)/nitr*1000000

timeit_M_pb = repeat("M_pb_func()", setup = "from __main__ import M_pb_func", repeat = nrepeat, number = nitr)
mediantime_M_pb = median(timeit_M_pb)/nitr*1000000

timeit_M_rbdl = repeat("M_rbdl_func()", setup = "from __main__ import M_rbdl_func", repeat = nrepeat, number = nitr)
mediantime_M_rbdl = median(timeit_M_rbdl)/nitr*1000000

timeit_M_kdl = repeat("M_kdl_func()", setup = "from __main__ import M_kdl_func", repeat = nrepeat, number = nitr)
mediantime_M_kdl = median(timeit_M_kdl)/nitr*1000000


aba_u2c = asd.get_forward_dynamics_aba(root, tip, gravity_u2c)
timeit_fdaba_u2c = repeat("aba_u2c_func()", setup = "from __main__ import aba_u2c_func", repeat = nrepeat, number = nitr)
mediantime_fdaba_u2c = median(timeit_fdaba_u2c)/nitr*1000000

crba_u2c = asd.get_forward_dynamics_crba(root, tip, gravity_u2c)
timeit_fdcrba_u2c = repeat("crba_u2c_func()", setup = "from __main__ import crba_u2c_func", repeat = nrepeat, number = nitr)
mediantime_fdcrba_u2c = median(timeit_fdcrba_u2c)/nitr*1000000

timeit_fd_rbdl = repeat("fd_rbdl_func()", setup = "from __main__ import fd_rbdl_func", repeat = nrepeat, number = nitr)
mediantime_fd_rbdl = median(timeit_fd_rbdl)/nitr*1000000


In [5]:
print "\npendulum_g_u2c=", mediantime_g_u2c
print "\npendulum_id_u2c=", mediantime_id_u2c
print "\npendulum_C_u2c=", mediantime_c_u2c
print "\npendulum_M_u2c=", mediantime_M_u2c

print "\npendulum_g_rbdl=", mediantime_g_rbdl
print "\npendulum_id_rbdl=", mediantime_id_rbdl
print "\npendulum_C_rbdl=", mediantime_c_rbdl
print "\npendulum_M_rbdl=", mediantime_M_rbdl

print "\npendulum_g_pb=", mediantime_g_pb
print "\npendulum_id_pb=", mediantime_id_pb
print "\npendulum_C_pb=", mediantime_c_pb
print "\npendulum_M_pb=", mediantime_M_pb

print "\npendulum_g_kdl=", mediantime_g_kdl
print "\npendulum_C_kdl=", mediantime_c_kdl
print "\npendulum_M_kdl=", mediantime_M_kdl

print "\nfdrbdl=", mediantime_fd_rbdl
print "\nfdcrba=", mediantime_fdcrba_u2c 
print "\nfdaba=", mediantime_fdaba_u2c


pendulum_g_u2c= 12.6429796219

pendulum_id_u2c= 28.8081169128

pendulum_C_u2c= 20.6830501556

pendulum_M_u2c= 13.1369829178

pendulum_g_rbdl= 2.99441814423

pendulum_id_rbdl= 5.16653060913

pendulum_C_rbdl= 3.65996360779

pendulum_M_rbdl= 2.39741802216

pendulum_g_pb= 2.32148170471

pendulum_id_pb= 4.18245792389

pendulum_C_pb= 3.22103500366

pendulum_M_pb= 2.27749347687

pendulum_g_kdl= 4.85944747925

pendulum_C_kdl= 6.12354278564

pendulum_M_kdl= 5.07247447968

fdrbdl= 5.3129196167

fdcrba= 28.9306640625

fdaba= 28.787612915



pendulum_g_u2c= 12.6429796219

pendulum_id_u2c= 28.8081169128

pendulum_C_u2c= 20.6830501556

pendulum_M_u2c= 13.1369829178

fdcrba= 28.9306640625

fdaba= 28.787612915

# Derivatives

## q

In [6]:
import casadi as cs

qs =cs.SX.sym("qs",njoints)
qsdot =cs.SX.sym("qsdot",njoints)
qsddot =cs.SX.sym("qsddot",njoints)
taus =cs.SX.sym("taus",njoints)

C_j = cs.jacobian(c_u2c(qs, qsdot), qs)
g_j = cs.jacobian(g_u2c(qs), qs)
M_j = cs.jacobian(M_u2c(qs), qs)
invd_j = cs.jacobian(ID_u2c(qs, qsdot, qsddot), qs)
fd_aba_j = cs.jacobian(aba_u2c(qs, qsdot, taus), qs)
fd_crba_j = cs.jacobian(crba_u2c(qs, qsdot, taus), qs)


g_j = cs.Function("g_j", [qs], [g_j], {"jit": True, "jit_options":{"flags":"-Ofast"}})
C_j = cs.Function("C_j", [qs, qsdot], [C_j], {"jit": True, "jit_options":{"flags":"-Ofast"}})
M_j = cs.Function("M_j", [qs], [M_j], {"jit": True, "jit_options":{"flags":"-Ofast"}})
invd_j = cs.Function("invd_j", [qs, qsdot, qsddot], [invd_j], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_aba_j = cs.Function("fd_aba_j", [qs, qsdot, taus], [fd_aba_j], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_crba_j = cs.Function("fd_crba_j", [qs, qsdot, taus], [fd_crba_j], {"jit": True, "jit_options":{"flags":"-Ofast"}})

In [7]:
def fdcrba_j_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    fd_crba_j(q_none, qdot_none, tau_none)

def fdaba_j_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    fd_aba_j(q_none, qdot_none, tau_none)
    
    
def M_j_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2   
    M_j(q_none)
    
def g_j_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    g_j(q_none)


def C_j_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    C_j(q_none, qdot_none)
    
def ID_j_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    invd_j(q_none, qdot_none, qddot_none)

def aba_j_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    fd_aba_j(q_none, qdot_none, tau_none)
    
def crba_j_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    
    fd_aba_j(q_none, qdot_none, tau_none)

In [8]:
timeit_idj = repeat("ID_j_func()", setup = "from __main__ import ID_j_func", repeat = nrepeat, number = nitr)
mediantime_idj = median(timeit_idj)/nitr*1000000

timeit_gj = repeat("g_j_func()", setup = "from __main__ import g_j_func", repeat = nrepeat, number = nitr)
mediantime_gj = median(timeit_gj)/nitr*1000000

timeit_Cj = repeat("C_j_func()", setup = "from __main__ import C_j_func", repeat = nrepeat, number = nitr)
mediantime_Cj = median(timeit_Cj)/nitr*1000000

timeit_Mj = repeat("M_j_func()", setup = "from __main__ import M_j_func", repeat = nrepeat, number = nitr)
mediantime_Mj = median(timeit_Mj)/nitr*1000000

timeit_fdabaj = repeat("aba_j_func()", setup = "from __main__ import aba_j_func", repeat = nrepeat, number = nitr)
mediantime_fd_abaj = median(timeit_fdabaj)/nitr*1000000

timeit_fdcrbaj = repeat("crba_j_func()", setup = "from __main__ import crba_j_func", repeat = nrepeat, number = nitr)
mediantime_fd_crbaj = median(timeit_fdcrbaj)/nitr*1000000

   
print "\npendulum_id_d=", mediantime_idj
print "\npendulum_g_d=", mediantime_gj
print "\npendulum_C_d=", mediantime_Cj
print "\npendulum_M_d=", mediantime_Mj
print "\npendulum_fdaba_d=", mediantime_fd_abaj
print "\npendulum_fdcrba_d=", mediantime_fd_crbaj


pendulum_id_d= 28.7179946899

pendulum_g_d= 12.5154256821

pendulum_C_d= 20.6383466721

pendulum_M_d= 12.5584602356

pendulum_fdaba_d= 28.6189317703

pendulum_fdcrba_d= 28.6304950714



pendulum_id_d= 28.7179946899

pendulum_g_d= 12.5154256821

pendulum_C_d= 20.6383466721

pendulum_M_d= 12.5584602356

pendulum_fdaba_d= 28.6189317703

pendulum_fdcrba_d= 28.6304950714


## qdot


In [9]:
qsdot =cs.SX.sym("qsdot",njoints)
qsddot =cs.SX.sym("qsddot",njoints)
taus =cs.SX.sym("taus",njoints)

C_jdot = cs.jacobian(c_u2c(qs, qsdot), qsdot)
invd_jdot = cs.jacobian(ID_u2c(qs, qsdot, qsddot), qsdot)
fd_aba_jdot = cs.jacobian(aba_u2c(qs, qsdot, taus), qsdot)
fd_crba_jdot = cs.jacobian(crba_u2c(qs, qsdot, taus), qsdot)


C_jdot = cs.Function("C_j", [qs, qsdot], [C_jdot], {"jit": True, "jit_options":{"flags":"-Ofast"}})
invd_jdot = cs.Function("invd_j", [qs, qsdot, qsddot], [invd_jdot], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_aba_jdot = cs.Function("fd_aba_j", [qs, qsdot, taus], [fd_aba_jdot], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_crba_jdot = cs.Function("fd_crba_j", [qs, qsdot, taus], [fd_crba_jdot], {"jit": True, "jit_options":{"flags":"-Ofast"}})

In [10]:
timeit_idj = repeat("ID_j_func()", setup = "from __main__ import ID_j_func", repeat = nrepeat, number = nitr)
mediantime_idj = median(timeit_idj)/nitr*1000000


timeit_Cj = repeat("C_j_func()", setup = "from __main__ import C_j_func", repeat = nrepeat, number = nitr)
mediantime_Cj = median(timeit_Cj)/nitr*1000000


timeit_fdabaj = repeat("aba_j_func()", setup = "from __main__ import aba_j_func", repeat = nrepeat, number = nitr)
mediantime_fd_abaj = median(timeit_fdabaj)/nitr*1000000

timeit_fdcrbaj = repeat("crba_j_func()", setup = "from __main__ import crba_j_func", repeat = nrepeat, number = nitr)
mediantime_fd_crbaj = median(timeit_fdcrbaj)/nitr*1000000

   
print "\npendulum_id_d=", mediantime_idj
print "\npendulum_C_d=", mediantime_Cj
print "\npendulum_fdaba_d=", mediantime_fd_abaj
print "\npendulum_fdcrba_d=", mediantime_fd_crbaj


pendulum_id_d= 28.9665460587

pendulum_C_d= 20.7270383835

pendulum_fdaba_d= 28.9844274521

pendulum_fdcrba_d= 28.8769006729



pendulum_id_d= 28.9665460587

pendulum_C_d= 20.7270383835

pendulum_fdaba_d= 28.9844274521

pendulum_fdcrba_d= 28.8769006729

## qddot / tau


In [11]:
qsdot =cs.SX.sym("qsdot",njoints)
qsddot =cs.SX.sym("qsddot",njoints)
taus =cs.SX.sym("taus",njoints)

invd_j = cs.jacobian(ID_u2c(qs, qsdot, qsddot), qsddot)
fd_aba_j = cs.jacobian(aba_u2c(qs, qsdot, taus), taus)
fd_crba_j = cs.jacobian(crba_u2c(qs, qsdot, taus), taus)

invd_j = cs.Function("invd_j", [qs, qsdot, qsddot], [invd_j], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_aba_j = cs.Function("fd_aba_j", [qs, qsdot, taus], [fd_aba_j], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_crba_j = cs.Function("fd_crba_j", [qs, qsdot, taus], [fd_crba_j], {"jit": True, "jit_options":{"flags":"-Ofast"}})

In [14]:
timeit_idj = repeat("ID_j_func()", setup = "from __main__ import ID_j_func", repeat = nrepeat, number = nitr)
mediantime_idj = median(timeit_idj)/nitr*1000000

timeit_fdabaj = repeat("aba_j_func()", setup = "from __main__ import aba_j_func", repeat = nrepeat, number = nitr)
mediantime_fd_abaj = median(timeit_fdabaj)/nitr*1000000

timeit_fdcrbaj = repeat("crba_j_func()", setup = "from __main__ import crba_j_func", repeat = nrepeat, number = nitr)
mediantime_fd_crbaj = median(timeit_fdcrbaj)/nitr*1000000

   
print "\npendulum_id_d=", mediantime_idj
print "\npendulum_fdaba_d=", mediantime_fd_abaj
print "\npendulum_fdcrba_d=", mediantime_fd_crbaj


ur5_id_d= 29.1345119476

ur5_fdaba_d= 30.308008194

ur5_fdcrba_d= 28.9859771729



pendulum_id_d= 29.1345119476

pendulum_fdaba_d= 28.879404068

pendulum_fdcrba_d= 28.7123918533
